rfcs/0126-standardize-magic-properties.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: standardize-magic-properties
+start-date: 2022-05-26
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Rename attrset attribute names treated specially by Nix to start with a double underscore.
+Add a builtin function `guardMagic` that returns `null` for double underscore prefixed strings.
+
+# Motivation
+[motivation]: #motivation
+
+Nix treats many attributes specially with unique behaviours. Some of these are prefixed with double underscores to disambiguate them from everything else. But some aren’t: `recurseForDerivations` isn’t, `type`, when set to `"derivation"`, changes visual output.
+
+It would also be a good idea to avoid accidentally using reserved special attribute names. Therefore I suggest `guardMagic` that can be used in dynamic attribute names to avoid this.
+
+# Detailed design
+[design]: #detailed-design
+
+Properties treated specially by the Nix tools and Nix language when dealing with arbitrary are renamed to start with a double underscore. Instances where this is not necessary are `__functor` and `__toString`. `recurseForDerivations` is renamed `__recurseForDerivations` and `type` is renamed `__type`.
+
+A new builtin function is introduced, `builtins.guardMagic`.
+If called on a string, it returns the string if it does not start with a double underscore, and `null` otherwise. This can be used by semantically charged non-generic functions that update attrs dynamically to avoid unwanted interactions.
+
+Nixpkgs & Nix need to be updated to reflect this. Therefore, it may be advisable to have a grace period during which unprefixed magic properties are still supported, but warn.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+```nix
+rec {
+  set-val-to-null = k: attrs: attrs // {
+    ${guardMagic k} = null;
+  };
+  x = set-val-to-null "__functor" { __functor = _: _: { }; };
+  y = set-val-to-null "y" { __functor = _: _: { }; };
+}
+```
+evaluates to
+```nix
+{
+  set-val-to-null = «lambda»;
+  x = { __functor = «lambda»; };
+  y = { y = null; __functor = «lambda»; };
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This change is backwards-incompatible
```

Comment:
I don't think it's such a good idea to make it actually breaking: IIRC one of the benefits Nix advertises is that you can instantiate old Nix expressions and still get the same result (but that also means that the behavior of the expression shouldn't change).

Reply:
I see, but then this reduces to just `guardMagic`.

Reply:
Also, this doesn’t stop old expressions from building, since `__toString` and `__functor` are the only ones here that aren’t user-facing, and they’re not renamed.

Reply:
When grepping for `recurseForDerivations` in `nixpkgs`, there's a bunch of places where it's explicitly used, so what do you mean with "user-facing"?

Reply:
Their special effect is outside the expression language. AFAIK `recurseForDerivations` tells nix commands to search for derivations deeper down, but it’s only being set in Nixpkgs, not used by it. Renaming it without renaming it in nixpkgs or using an old nix expression will not affect directly building any particular derivation, but only auxiliary things like searching. It’s unlikely that many nix expression directly check `recurseForDerivations`.
The other ones, on the other hand, if changed, would actually stop some Nix expressions from evaluating. `__functor` concerns the in-expression behaviour of these values.

Reply:
Technically speaking, _adding_ synonyms that start with `__` for all the magic attributes and deciding that Nixpkgs will use the underscore-based things from now on is more than `guardMagic`.

Evaluation-affecting magic seems to be indeed underscored; `nix` command behaviour is nominally still experimental so some changes are possible with proper warning.

Reply:
> Evaluation-affecting magic seems to be indeed underscored; nix command behaviour is nominally still experimental so some changes are possible with proper warning.

But doesn't that affect nix-* commands usually as well?

Reply:
It does. In my opinion that’s not as much a breaking change as a change in how some *derivations* build.